### PR TITLE
Fix for #231 (FtRemote hides TkFork response status)

### PR DESCRIPTION
**Pull request by cyberone**
Fix for https://github.com/yegor256/takes/issues/231 (FtRemote hides TkFork response status)

1. Wrapped failure handling at BkBasic with response code depending on exception class
1. Unit-test for response code handling

### CODE REVIEW

**Comment by davvd** [comment]
@cyberone Let me find a reviewer for this pull request, thanks for submitting it

**Comment by davvd** [comment]
@carlosmiranda this pull request is for you, please review

File: src/main/java/org/takes/http/BkBasic.java [hunk 1]
```diff
@@ -116,10 +117,17 @@ private static Response failure(final Throwable err) {
         writer.close();
         return new RsWithStatus(
             new RsText(new ByteArrayInputStream(baos.toByteArray())),
-            HttpURLConnection.HTTP_INTERNAL_ERROR
+            responseCode(err)
         );
     }
 
+    private static int responseCode(Throwable err) {
```
Comment (carlosmiranda) [inline src/main/java/org/takes/http/BkBasic.java]: @cyberone Where is the Javadoc comment for this method?

File: src/main/java/org/takes/http/BkBasic.java [hunk 2]
```diff
@@ -116,10 +117,17 @@ private static Response failure(final Throwable err) {
         writer.close();
         return new RsWithStatus(
             new RsText(new ByteArrayInputStream(baos.toByteArray())),
-            HttpURLConnection.HTTP_INTERNAL_ERROR
+            responseCode(err)
```
Comment (carlosmiranda) [inline src/main/java/org/takes/http/BkBasic.java]: @cyberone Let's add Class prefix here (i.e. `BkBasic.responseCode(err)`).

File: src/test/java/org/takes/http/BkBasicTest.java [hunk 3]
```diff
@@ -77,4 +86,26 @@ public void handlesSocket() throws IOException {
             Matchers.containsString("Hello world")
         );
     }
+
+    /**
+     * BkBasic can return HTTP status 404 when accessing invalid URL.
+     * @throws IOException if any I/O error occurs.
+     */
+    @Test
+    public void returns404OnInvalidUrl() throws IOException {
```
Comment (carlosmiranda) [inline src/test/java/org/takes/http/BkBasicTest.java]: @cyberone Let's declare `throws Exception` as per our [code standards](http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names).

File: src/test/java/org/takes/http/BkBasicTest.java [hunk 4]
```diff
@@ -27,12 +27,21 @@
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.net.HttpURLConnection;
 import java.net.InetAddress;
 import java.net.Socket;
+import java.net.URI;
+
```
Comment (carlosmiranda) [inline src/test/java/org/takes/http/BkBasicTest.java]: @cyberone Checkstyle will complain about this extra space, and I think import order is wrong (should be alphabetic).

**Comment by carlosmiranda** [comment]
@cyberone See my comments above. There are some Checkstyle errors also, please fix them.

**Comment by cyberone** [comment]
@carlosmiranda I'm sorry, qulice was turned off. Check again please.

**Comment by carlosmiranda** [comment]
@rultor ready to merge

File: src/main/java/org/takes/http/BkBasic.java [hunk 5]
```diff
@@ -116,11 +117,24 @@ private static Response failure(final Throwable err) {
...
     }
 
     /**
+     * Returns HTTP response code for specified exception.
+     * @param err Exception
+     * @return Numeric HTTP code.
+     */
+    private static int responseCode(final Throwable err) {
+        int result = HttpURLConnection.HTTP_INTERNAL_ERROR;
+        if (err instanceof HttpException) {
```
Comment (yegor256) [inline src/main/java/org/takes/http/BkBasic.java]: we don't use `instanceof`, it's evil, see http://www.yegor256.com/2015/04/02/class-casting-is-anti-pattern.html

**Comment by yegor256** [comment]
@cyberone see above

**Comment by cyberone** [comment]
@yegor256 done, check again please. Is there a list of all teamed.io's coding standarts, patterns and anti-patterns?

**Comment by yegor256** [comment]
@rultor merge

**Comment by yegor256** [comment]
@cyberone well, yes, here is a good starting point: http://www.yegor256.com/2014/09/10/anti-patterns-in-oop.html

**Comment by davvd** [comment]
@elenavolokhova please, let us know what do you think about this ticket, according to our [QA rules](http://www.teamed.io/qa.html)

**Comment by davvd** [comment]
@rultor please deploy

**Comment by elenavolokhova** [comment]
@davvd It looks good!

**Comment by davvd** [comment]
> @davvd It looks good!

@elenavolokhova thanks a lot :)

**Comment by davvd** [comment]
@carlosmiranda 10 mins was added to the account of @elenavolokhova (for QA review), in transaction 57291301; Thank you, I have added **17 mins** to your account in payment/transaction "`57291305`", time consumed: 119 hours and 52 mins; you're getting extra minutes here (c=2); added +17 to your rating, now it is equal to [+4223](http://www.netbout.com/b/35023?open=rating)

**Pull request merged**